File: app/enums.cpp
#include "enums.h"

#include <QDataStream>

ElementType Enums::nextElmType(ElementType type) {
    switch (type) {
        // Not and Node
        case ElementType::Not:
            return ElementType::Node;
        case ElementType::Node:
            return ElementType::Not;

        // Gates
        case ElementType::And:
            return ElementType::Or;
        case ElementType::Or:
            return ElementType::Nand;
        case ElementType::Nand:
            return ElementType::Nor;
        case ElementType::Nor:
            return ElementType::Xor;
        case ElementType::Xor:
            return ElementType::Xnor;
        case ElementType::Xnor:
            return ElementType::And;

        // StaticInput and Input
        case ElementType::InputVcc:
            return ElementType::InputGnd;
        case ElementType::InputGnd:
            return ElementType::InputButton;
        case ElementType::InputButton:
            return ElementType::InputSwitch;
        case ElementType::InputSwitch:
            return ElementType::InputRotary;
        case ElementType::InputRotary:
            return ElementType::Clock;
        case ElementType::Clock:
            return ElementType::InputVcc;

        // FipFlops
        case ElementType::DFlipFlop:
            return ElementType::TFlipFlop;
        case ElementType::TFlipFlop:
            return ElementType::DFlipFlop;

        case ElementType::JKFlipFlop:
            return ElementType::SRFlipFlop;
        case ElementType::SRFlipFlop:
            return ElementType::JKFlipFlop;

        // Output
        case ElementType::Buzzer:
            return ElementType::Led;
        case ElementType::Led:
            return ElementType::Buzzer;

        // Other Cases (falltrought)
        default:
            return ElementType::Unknown;
    }
}

ElementType Enums::prevElmType(ElementType type) {
    switch (type) {
        // Not and Node
        case ElementType::Not:
            return ElementType::Node;
        case ElementType::Node:
            return ElementType::Not;

        // Gates
        case ElementType::And:
            return ElementType::Xnor;
        case ElementType::Xnor:
            return ElementType::Xor;
        case ElementType::Xor:
            return ElementType::Nor;
        case ElementType::Nor:
            return ElementType::Nand;
        case ElementType::Nand:
            return ElementType::Or;
        case ElementType::Or:
            return ElementType::And;

        // StaticInput and Input
        case ElementType::InputVcc:
            return ElementType::Clock;
        case ElementType::Clock:
            return ElementType::InputRotary;
        case ElementType::InputRotary:
            return ElementType::InputSwitch;
        case ElementType::InputSwitch:
            return ElementType::InputButton;
        case ElementType::InputButton:
            return ElementType::InputGnd;
        case ElementType::InputGnd:
            return ElementType::InputVcc;

        // FipFlops
        case ElementType::DFlipFlop:
            return ElementType::TFlipFlop;
        case ElementType::TFlipFlop:
            return ElementType::DFlipFlop;

        case ElementType::JKFlipFlop:
            return ElementType::SRFlipFlop;
        case ElementType::SRFlipFlop:
            return ElementType::JKFlipFlop;

        // Output
        case ElementType::Led:
            return ElementType::Buzzer;
        case ElementType::Buzzer:
            return ElementType::Led;

        // Other Cases (falltrought)
        default:
            return ElementType::Unknown;
    }
}

ElementType &operator++(ElementType &type)
{
    return type = static_cast<ElementType>(static_cast<int>(type) + 1);
}

QDataStream &operator>>(QDataStream &stream, ElementType &type)
{
    quint64 temp; stream >> temp;
    type = static_cast<ElementType>(temp);
    return stream;
}

QDataStream &operator<<(QDataStream &stream, const ElementType &type)
{
    stream << static_cast<quint64>(type);
    return stream;
}

File: mcp/client/domain/enums.py
#!/usr/bin/env python3
"""
Domain Enumerations

This module contains enumerations that define the domain concepts
for circuit simulation, element types, and export formats.
"""

from enum import Enum


class SimulationAction(str, Enum):
    """Valid simulation control actions"""

    START = "start"
    STOP = "stop"
    RESTART = "restart"
    UPDATE = "update"


class ElementType(str, Enum):
    """Valid element types for circuit creation"""

    # Logic Gates
    AND = "And"
    OR = "Or"
    NOT = "Not"
    NOR = "Nor"
    NAND = "Nand"
    XOR = "Xor"
    XNOR = "Xnor"
    NODE = "Node"

    # Inputs
    INPUT_BUTTON = "InputButton"
    INPUT_SWITCH = "InputSwitch"
    INPUT_ROTARY = "InputRotary"
    INPUT_VCC = "InputVcc"
    INPUT_GND = "InputGnd"
    CLOCK = "Clock"

    # Outputs
    LED = "Led"
    BUZZER = "Buzzer"
    AUDIO_BOX = "AudioBox"
    DISPLAY_7 = "Display7"
    DISPLAY_14 = "Display14"
    DISPLAY_16 = "Display16"

    # Memory/Sequential Logic
    DFLIPFLOP = "DFlipFlop"
    DLATCH = "DLatch"
    JKFLIPFLOP = "JKFlipFlop"
    SRFLIPFLOP = "SRFlipFlop"
    SRLATCH = "SRLatch"
    TFLIPFLOP = "TFlipFlop"

    # Multiplexers
    MUX = "Mux"
    DEMUX = "Demux"

    # Special Elements
    IC = "IC"
    TRUTH_TABLE = "TruthTable"
    LINE = "Line"
    TEXT = "Text"


class ImageExportFormat(str, Enum):
    """Valid export formats for circuit images"""

    PNG = "png"
    SVG = "svg"
    PDF = "pdf"


class WaveformExportFormat(str, Enum):
    """Valid export formats for waveform data and charts"""

    TXT = "txt"
    PNG = "png"
